### PR TITLE
Explicit handling of a different Executor in TransportAction

### PR DESCRIPTION
**Pull request by ldematte**
Relates to https://github.com/elastic/elasticsearch/issues/97916

Introduces a Executor parameter (and field) to `TransportAction`, and updates its `execute()` method to fork using the executor if needed.
An additional method `executeDirect` is added so inheritors can have the possibility to bypass forking and keep running on the current executor, e.g. if they have already (e.g.  `HandledTransportAction`).
Clients like NodeClient will then do not need to worry about forking, as it will be done by `TransportAction` directly when `execute()` is called.

Provisionally all classes are updated to accept the direct executor, so behaviour is largerly unchanged (except of `HandledTransportAction`, where we already propagate the executor).
A follow up step would be to revise  `TransportAction` implementations overriding `doExecute` to fork on a different executor (in order to work around https://github.com/elastic/elasticsearch/issues/97916) and simplify them.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
Seems reasonable to me, although I think it deserves some tests to check that we do/don't fork as expected. I'd also like Ryan to take a look.

**Comment by ldematte** [comment]
> I think it deserves some tests to check that we do/don't fork as expected
++ 
I wanted first to check with you the approach was reasonable. I'm adding Ryan too.

**Comment by ldematte** [comment]
@elasticsearchmachine update branch

**Comment by ldematte** [comment]
@elasticmachine update branch

**Comment by ldematte** [comment]
@elasticmachine rerun elasticsearch-ci/bwc

**Review by rjernst — COMMENTED** [review]
tl;dr I'm hesitant about this change, but I don't feel strongly enough to push it one way or another.

Over the last many years I've worked to reduce the number of required arguments to TransportAction. The reason was to make transport actions "dumber". That is, I wanted to make them easier to construct (at least the base class) to reduce the number of extra things that need to be passed around. There are still some left (specifically the task manager and action filters), but I have thoughts on how to remove those as well, it's just quite invasive so will require some more thought.

Ultimately my goal was to try to get rid of the construction of transport actions directly through the guice injector, and instead have all transport actions tied to services. Effectively every service could present some set of rest actions and transport actions that are bound to methods on that service. Then any dependencies of running that actions become just dependencies of the service, rather than the action directly. I think that would create an easier to understand object dependency tree.

With this change, executors are added directly to the TransportAction ctor args. I can see how it is useful, and it makes sense that the action definition is tied to a desired executor. Perhaps it can even fit into the world I described above, where each service sets the executor when constructing the lightweight binding to its own service method. I'm not so keen on the "transport action handler" terminology, but I think that can be removed in the future with some more action filter refactoring.

To summarize, I'm not sure I like this, but I don't feel strongly enough to block it. If others feel the benefits are large enough, then let's proceed, and we can work out separately how to make transport action construction lightweight.

**Comment by ldematte** [comment]
tl;dr: I'm not attached to this just because I created this PR: if you think it might not be a step forward but instead could complicate things for future improvements, let's ditch this and think about other solutions.

I understand why you might be reluctant to accept this; from what you wrote, it goes against the direction you have worked on so far. Even if maybe not strongly - as you pointed out, it kind of make sense to tie an action to a preferred executor - it does not go in the same direction.
In case you missed it, there is a long-ish discussion on the issue opened by David: https://github.com/elastic/elasticsearch/issues/97916
As you can see, we agree that this solution is a compromise. Maybe one of the other options is more aligned with your plans. For example, one of the ideas was to act on the registration side - perhaps that could help more in the direction of getting rid of the guice injector.
Another more radical idea (by @DaveCTurner - see https://github.com/elastic/elasticsearch/issues/100111 - I definitely cannot get the credit for it!) is to completely separate "local actions" (only invoked on the local node by a local client (NodeClient))  from transport actions (only invoked via the transport service), using different types. The new "local actions" will be created or registered with an executor; NodeClient will know about it and act accordingly. Being a new (set of) types, TransportAction will not be touched.

**Comment by rjernst** [comment]
> completely separate "local actions" (only invoked on the local node by a local client (NodeClient)) from transport actions (only invoked via the transport service), using different types

My quick thoughts on this are that local actions, combined with executor at registration (for the non-local actions) would be better aligned with de-guicifying transport actions. Forcing an executor in the type hierarchy in this PR would mean a dummy executor is needed for local actions, which doesn't seem right.

**Comment by ldematte** [comment]
> My quick thoughts on this are that local actions, combined with executor at registration (for the non-local actions) would be better aligned with de-guicifying transport actions.

This I understand and looks appealing. In practice, it will promote/normalize what is "hidden" inside `HandledTransportAction` today (`transportService.registerRequestHandler(... executor ...)`.
But I don't get why 

> Forcing an executor in the type hierarchy in this PR would mean a dummy executor is needed for local actions, which doesn't seem right.

This I do not understand.
Local actions will need an executor anyway, right? The whole effort is to provide a way for locally-executed actions (via NodeClient) to fork and run off the calling thread.

**Comment by DaveCTurner** [comment]
> Forcing an executor in the type hierarchy in this PR would mean a dummy executor is needed for local actions, which doesn't seem right.

I think there's a misunderstanding here. This PR is adding the executor specifically for local actions to use. Non-local actions already dispatch their work to the executor with which they were registered with the `TransportService`.

Possibly we should be passing an executor to the `ActionHandler` with which the action is registered instead? We already have created the `ThreadPool` at this point, we could add some plumbing to make it available at registration here.

**Comment by ldematte** [comment]
> Possibly we should be passing an executor to the ActionHandler with which the action is registered instead? We already have created the ThreadPool at this point, we could add some plumbing to make it available at registration here.

That is an interesting (and more lightweight?) alternative to the 3rd option here: https://github.com/elastic/elasticsearch/issues/97916#issuecomment-1750211977. Instead of changing how the action is registered, we can change how the action (ActionHandler) is declared. Then the registration code can extract the additional information (the Executor) and build a richer Map for NodeClient (or 2 maps).

**Comment by DaveCTurner** [comment]
Yeah it's in a similar direction at least.

The thing I don't particularly like about this idea is that, as things are written today, the `ActionModule` registration happens (and therefore the executor would be declared) a long way from the code that the action executes. That's different from the `TransportService` registration which happens during construction of the `TransportAction`. I would prefer all this information to be in the same source file as far as possible so that threading bugs are more likely to be caught in reviews.

I can see some advantages with achieving this by moving the `ActionModule` registration into the action's construction path alongside the `TransportService` registration. It's kind of weird that they're so different when they're basically doing the same things. But that's going to be quite a substantial change and maybe not one we are ready to tackle just yet. IMO the approach here is not a terrible alternative in the interim, and in particular I think it doesn't really stop us from improving the registration path later on.

(is that enough to move Ryan from a -0 to a +0 at least?)

**Comment by ldematte** [comment]
> It's kind of weird that they're so different when they're basically doing the same things.

Oh yes, it took me a while to figure out how and where registration happens, and I'm sure I'm still missing bits of it.
And I agree they are "far from home", in a place where people can forget them. But having all the information in one source file/one class seems at odd with the idea of making them as lightweight as possible.
My personal preference would be towards a single, well defined and visible place to register all actions, but external to the action itself (the current `HandledTransportAction` pattern is not one I am fond of to be honest) because I like simple constructors, small classes and linear flows. But I'm not dogmatic :) and I trust your opinions since you have so much more experience in the codebase.

**Comment by DaveCTurner** [comment]
> placing it on the top of the TransportAction hierarchy forces non-local actions to specify an executor.

Indeed. But non-local actions already need to decide on an executor for their `TransportService` registration, so this isn't a particularly unnatural thing to put here for now.

> All of this is to say, I think we should be looking away from registration within ctors, not adding to it.

I very much agree - I'd like to see us do neither registration in the ctor, but I would hope we can get to a place where both registrations happen in basically the same way somewhere fairly close to each other (in the source code, if not in program flow) as part of constructing all the actions and their dependent services. That's what I meant by the phrase "construction path", intended to cover more things than just calling the ctor itself.

**Comment by ldematte** [comment]
Since it seems we all agree this is "not bad" (+0), but we'd all like an approach where a) both registrations happen in basically the same way and b) registration happens outside the constructor (but on the action construction path), I have a proposal: why don't we put this PR on hold, and in my spare time I try a different approach based on what we discussed here? It may end up nowhere, if I bump into things that are too complex to disentangle as they are now, but it will at least give us a comparison. What do you think?

**Comment by DaveCTurner** [comment]
I'm a little more positive on this change than +0, I think it is a reasonable step towards addressing https://github.com/elastic/elasticsearch/issues/97916 which is a source of genuine bugs today. We have put some workarounds in place for the worst offenders but they're kinda ugly and still do the request-deserialization work on the wrong thread (since we consider dispatching to the executor twice would be worse).

I'm also +1 on looking for better alternatives first. This change isn't really a step in the right direction, although it's not really a step backwards either.

**Comment by ldematte** [comment]
@DaveCTurner @rjernst I have a (very preliminary) PoC here: https://github.com/ldematte/elasticsearch/pull/1
When you have some time, can you please give it a look and let me know if you feel it's a more promising alternative?

**Comment by ldematte** [comment]
Closing this as it seems we want to go in a different direction

**Comment by ldematte** [comment]
Re-opening: 
- we have a clearer idea of what the "different direction" will be: we still will have DI, but it will be properly tailored to our needs and will be used to sort out TransportAction creation and registration.
   - this is worked on as our effort to reduce technical debt, but it's still some months away
- meanwhile, we had at least one incident which had https://github.com/elastic/elasticsearch/issues/97916 as root cause again, and we have an action item to mitigate the risk of this happening again (see https://elasticco.atlassian.net/browse/ES-8624)
- after syncing with the team, we agreed to proceed with this as a temporary stopgap measure: we *still* want to solve this properly, in a way that will likely undo this change, but we want to be more protected meanwhile.

**Review by rjernst — APPROVED** [review]
LGTM
